### PR TITLE
Design and implement list flyweight using variant, enum and union fields

### PR DESCRIPTION
**Pull request by izzymoriguchi**
Designed a flyweight class for a `list` that contains `variant`, `enum`and `union` fields. 

#### Possible types for `list` fields that are used in AMQP's `protocol.idl` are the following:
- `variant` of following types
    - string, string32
    - uint64, uint32, uint16, uint8, int64
    - array
    - map
    - octets
- Special case of `variant` where it skips a kind with `+`
- `variant` that has consecutive case statements
- `typedef` 
- `enum`
- `union`

### CODE REVIEW

File: src/test/java/org/reaktivity/nukleus/maven/plugin/internal/generated/ListWithUnionFW.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            switch (field)
+            {
+            case FIELD_INDEX_UNION_OCTETS:
+                if ((bitmask & (1 << FIELD_INDEX_UNION_OCTETS)) != 0)
+                {
+                    if (null == unionOctetsRO.tryWrap(buffer, fieldLimit, maxLimit))
+                    {
+                        return null;
+                    }
+                    fieldLimit = unionOctetsRO.limit();
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/maven/plugin/internal/generated/ListWithUnionFW.java]:
```
final UnionOctetsFW unionOctets = unionOctetsRO.tryWrap(buffer, fieldLimit, maxLimit);
if (unionOctets == null)
{
    return null;
}
fieldLimit = unionOctets.limit();
```

File: src/test/java/org/reaktivity/nukleus/maven/plugin/internal/generated/ListWithUnionFW.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    @Override
+    public String toString()
+    {
+        final long bitmask = bitmask();
+        return String.format("LIST_WITH_UNION [bitmask=%s%s, field1=%d]",
+            String.format("0x%04X", bitmask),
+            (bitmask & (1 << FIELD_INDEX_UNION_OCTETS)) != 0 ? String.format(", unionOctets=%s", unionOctets()) : "",
+            field1());
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/maven/plugin/internal/generated/ListWithUnionFW.java]: Effective bitmask includes defaulted fields, so ORing explicit bitmask with defaults mask gives effective bitmask for fields to be included in `toString()`.

Comment (jfallows) [reply src/test/java/org/reaktivity/nukleus/maven/plugin/internal/generated/ListWithUnionFW.java]:
```
boolean unionOctetsIsSet = (bitmask & (1 << FIELD_INDEX_UNION_OCTETS)) != 0;

StringBuilder format = new StringBuilder();
format.append("LIST_WITH_UNION [bitmask=%s");
if (unionOctetsIsSet)
{
    format.append(", unionOctets=%s");
}
...
format.append("]");

return MessageFormat.format(format.toString(), unionOctetsIsSet ? unionOctets() : null, ...);
```

File: src/main/antlr4/org/reaktivity/nukleus/maven/plugin/internal/parser/Nukleus.g4 [hunk 3]
```diff
@@ -256,7 +257,38 @@ array_member
...
+list_length
+   : LEFT_SQUARE_BRACKET unsigned_integer_type COMMA unsigned_integer_type (COMMA uint_literal)? RIGHT_SQUARE_BRACKET
+   ;
+
+list_members
+   : list_member* list_unbounded_member?
+   ;
+
+list_member
+   : KW_REQUIRED? type_spec declarators SEMICOLON
```
[the remark this answers is not in the document]
Comment (izzymoriguchi) [reply src/main/antlr4/org/reaktivity/nukleus/maven/plugin/internal/parser/Nukleus.g4]: Could you elaborate on this question/feedback?

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/ScopeVisitor.java [hunk 4]
```diff
@@ -217,7 +215,9 @@ private int findTypeId(
         AstStructNode currentNode = structNode;
         while (currentNode != null && currentNode.typeId() == 0 && currentNode.supertype() != null)
         {
-            currentNode = (AstStructNode) resolver.resolve(currentNode.supertype().name());
+            AstNamedNode namedNode = resolver.resolve(currentNode.supertype().name());
+
+            currentNode = namedNode.getKind().equals(Kind.STRUCT) ? (AstStructNode) namedNode : null;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/ScopeVisitor.java]: `==` is okay for enums, no need for `.equals(...)`

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/StructVisitor.java [hunk 5]
```diff
@@ -53,37 +55,36 @@ public StructVisitor(
 
     @Override
     public Collection<TypeSpecGenerator<?>> visitStruct(
-        AstNamedNode namedNode)
+        AstStructNode namedNode)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/StructVisitor.java]: `namedNode` -> `structNode`

File: src/main/antlr4/org/reaktivity/nukleus/maven/plugin/internal/parser/Nukleus.g4 [hunk 6]
```diff
@@ -278,9 +278,9 @@ list_members
 list_member
    : KW_REQUIRED? type_spec declarators SEMICOLON
    | KW_REQUIRED? uint_member_with_default SEMICOLON
-   | KW_REQUIRED? int_member_with_default SEMICOLON
-   | KW_REQUIRED? octets_member_with_default SEMICOLON
-   | KW_REQUIRED? non_primitive_member_with_default SEMICOLON
+   | int_member_with_default SEMICOLON
+   | octets_member_with_default SEMICOLON
+   | non_primitive_member_with_default SEMICOLON
```
[the remark this answers is not in the document]
Comment (izzymoriguchi) [reply src/main/antlr4/org/reaktivity/nukleus/maven/plugin/internal/parser/Nukleus.g4]: Oh sorry I just missed removing `required` keyword from that one line.

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/Parser.java [hunk 7]
```diff
@@ -155,15 +156,15 @@ private AstSpecificationNode parseSpecification(
 
         @Override
         public Set<String> visitStruct(
-            AstStructNode structNode)
+            AstStructNode namedNode)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/Parser.java]: `namedNode` -> `structNode`

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/StructVisitor.java (hunk 5 above)
Comment (jfallows) [reply src/main/java/org/reaktivity/nukleus/maven/plugin/internal/ast/visit/StructVisitor.java]: Still needs to be renamed, right?

**Pull request merged**